File: Java/Problem_057.java
import java.util.*;

public class Euler57 {

    static ArrayList<Integer> addDigits(List<Integer> a, List<Integer> b) {
        int n = Math.max(a.size(), b.size());
        ArrayList<Integer> out = new ArrayList<>(n + 2);
        int carry = 0;

        for (int i = 0; i < n; i++) {
            int da = (i < a.size()) ? a.get(i) : 0;
            int db = (i < b.size()) ? b.get(i) : 0;
            int s = da + db + carry;
            out.add(s % 10);
            carry = s / 10;
        }
        while (carry > 0) {
            out.add(carry % 10);
            carry /= 10;
        }
        return out;
    }

    static ArrayList<Integer> addScaledDigits(List<Integer> a, List<Integer> b, int k) {
        int n = Math.max(a.size(), b.size());
        ArrayList<Integer> out = new ArrayList<>(n + 3);
        int carry = 0;

        for (int i = 0; i < n; i++) {
            int da = (i < a.size()) ? a.get(i) : 0;
            int db = (i < b.size()) ? b.get(i) : 0;
            int s = da + k * db + carry;
            out.add(s % 10);
            carry = s / 10;
        }
        while (carry > 0) {
            out.add(carry % 10);
            carry /= 10;
        }
        return out;
    }

    static int solve() {
        ArrayList<Integer> p = new ArrayList<>();
        ArrayList<Integer> q = new ArrayList<>();
        p.add(3); // 3
        q.add(2); // 2

        int count = 0;

        for (int i = 0; i < 1000; i++) {
            if (p.size() > q.size()) count++;

            ArrayList<Integer> newP = addScaledDigits(p, q, 2); // p + 2q
            ArrayList<Integer> newQ = addDigits(p, q);          // p + q
            p = newP;
            q = newQ;
        }
        return count;
    }

    public static void main(String[] args) {
        System.out.println(solve()); // 153
    }
}

File: Python/Problem_057.py
def add_digits(a: list[int], b: list[int]) -> list[int]:
    n = max(len(a), len(b))
    out = []
    carry = 0
    for i in range(n):
        da = a[i] if i < len(a) else 0
        db = b[i] if i < len(b) else 0
        s = da + db + carry
        out.append(s % 10)
        carry = s // 10
    while carry:
        out.append(carry % 10)
        carry //= 10
    return out

def add_scaled_digits(a: list[int], b: list[int], k: int) -> list[int]:
    n = max(len(a), len(b))
    out = []
    carry = 0
    for i in range(n):
        da = a[i] if i < len(a) else 0
        db = b[i] if i < len(b) else 0
        s = da + k * db + carry
        out.append(s % 10)
        carry = s // 10
    while carry:
        out.append(carry % 10)
        carry //= 10
    return out

def euler_57() -> int:
    p = [3]  # 3
    q = [2]  # 2
    count = 0

    for _ in range(1000):
        if len(p) > len(q):
            count += 1
        new_p = add_scaled_digits(p, q, 2)  # p + 2q
        new_q = add_digits(p, q)            # p + q
        p, q = new_p, new_q

    return count

if __name__ == "__main__":
    print(euler_57())  # 153
